orderer/common/cluster/block_fetcher_test.go

```diff
@@ -355,6 +391,67 @@ func TestBlockFetcherBFTBehaviorBlockWithhold(t *testing.T) {
 	require.Equal(t, uint64(1), block.Header.Number)
 }
 
+func TestBlockFetcherAttestationSourceError(t *testing.T) {
+	// The first endpoint we try to pull Blocks witholds blocks,
+	// the PullBlock should then suspect the endpoint and probe other endpoints to confirm the suspicion.
+	// But attestation source factory fails to create attestation sources, probing should fail and
+	// no block is pulled.
+	bf := cluster.BlockFetcher{}
+
+	getNameForSource := nameFactory("node", 10)
+
+	bf.BlockSourceFactory = func(c cluster.FetcherConfig, latestConfigBlock *common.Block) (cluster.BlockSource, error) {
+		// node1 witholds block while other endpoints can deliver blocks
+		node_name := getNameForSource()
+		if node_name == "node1" {
+			return mock_block_puller_returns_nil(time.Millisecond * 4), nil
+		}
+		// the block puller below doesn't withold blocks
+		data := [][]byte{[]byte(node_name)}
+		// this block puller returns blocks with seq:1 and data after 2s
+		return mock_block_puller(1, data, time.Second*2), nil
+	}
+
+	attestation_source_created := false
+	lock := sync.Mutex{}
+	bf.AttestationSourceFactory = func(c cluster.FetcherConfig, latestConfigBlock *common.Block) (cluster.AttestationSource, error) {
+		lock.Lock()
+		defer lock.Unlock()
+		if !attestation_source_created {
+			// first attestation source created withholds attestation
+			attestation_source_created = true
+			return mock_attestation_puller_returns_nil(nil, time.Millisecond*12), nil
+		}
+		// all other attestation source fail to be created
+		return mock_attestation_puller(1, time.Second*2), errors.New("failed to pull attestations")
+	}
+
+	// Add time
+	bf.TimeNow = time.Now
+	bf.MaxRetries = 3
+	// Disable shuffle timeout
+	bf.CensorshipSuspicionThreshold = time.Duration(1 * time.Millisecond)
+	bf.PeriodicalShuffleInterval = time.Duration(1 * time.Hour)
+	bf.MaxByzantineNodes = 3
+	bf.Logger = flogging.MustGetLogger("test")
+	bf.FetcherConfig.FetchTimeout = time.Duration(time.Millisecond * 3)
+
+	bf.FetcherConfig.Endpoints = []cluster.EndpointCriteria{
+		{Endpoint: "localhost:5100"}, {Endpoint: "localhost:5101"}, {Endpoint: "localhost:5102"}, {Endpoint: "localhost:5103"}, {Endpoint: "localhost:5104"}, {Endpoint: "localhost:5105"}, {Endpoint: "localhost:5106"}, {Endpoint: "localhost:5107"}, {Endpoint: "localhost:5108"},
+	}
+	bf.BlockVerifierFactory = func(block *common.Block) cluster.BlockVerifierFunc {
+		return func(header *common.BlockHeader, metadata *common.BlockMetadata) error {
+			// assuming blocks are valid
+			return nil
+		}
+	}
+
+	bf.VerifyBlock = bf.BlockVerifierFactory(nil)
+
+	block := bf.PullBlock(1)
+	require.Equal(t, (*common.Block)(nil), block)
```

Comment:
once again - how do you know why you got nil? You might have gotten nil because of a different reason than you expect